src/Compilers/VisualBasic/Portable/Compilation/VisualBasicDeterministicKeyBuilder.vb

```diff
@@ -0,0 +1,75 @@
+﻿' Licensed to the .NET Foundation under one or more agreements.
+' The .NET Foundation licenses this file to you under the MIT license.
+' See the LICENSE file in the project root for more information.
+
+Namespace Microsoft.CodeAnalysis.VisualBasic
+
+    Friend NotInheritable Class VisualBasicDeterministicKeyBuilder
+        Inherits DeterministicKeyBuilder
+
+        Public Sub New()
+
+        End Sub
+
+        Protected Overrides Sub WriteParseOptionsCore(writer As JsonWriter, parseOptions As ParseOptions)
+            ' This can happen for SyntaxTree that are constructed via the API. 
+            If parseOptions Is Nothing Then
+                Return
+            End If
+
+            Dim basicOptions = TryCast(parseOptions, VisualBasicParseOptions)
+            If basicOptions Is Nothing Then
+                Throw New InvalidOperationException()
+            End If
+
+            MyBase.WriteParseOptionsCore(writer, parseOptions)
+
+            writer.Write("languageVersion", basicOptions.LanguageVersion)
+            writer.Write("specifiedLanguageVersion", basicOptions.SpecifiedLanguageVersion)
+
+            If basicOptions.PreprocessorSymbols.Length > 0 Then
+                writer.WriteKey("preprocessorSymbols")
+                writer.WriteArrayStart()
+                For Each pair In basicOptions.PreprocessorSymbols
+                    writer.WriteObjectStart()
+                    writer.WriteKey(pair.Key)
+                    writer.Write(pair.Value.ToString())
```

Comment:
interesting.  this is .ToString on an aribtrary object.  Do we know what forms this will take and if this is safe?

Reply:
this enumerates a dictionary.  that may have different results across rules of hte same program as i don't believe .net guarantees deterministic order.  Consider sorting things like this to ensure we're ok.